vllm/v1/sample/rejection_sampler.py

```diff
@@ -0,0 +1,153 @@
+# SPDX-License-Identifier: Apache-2.0
+import torch
+import torch.nn as nn
+
+from vllm.logger import init_logger
+from vllm.v1.outputs import SamplerOutput
+from vllm.v1.sample.metadata import SamplingMetadata
+
+try:
+    import flashinfer.sampling as fs
+    is_flashinfer_available = True
+except ImportError:
+    is_flashinfer_available = False
+
+logger = init_logger(__name__)
+INVALID_TOKEN_ID = -1
+
+
+class RejectionSampler(nn.Module):
+
+    def sample(self, logits: torch.Tensor,
+               sampling_metadata: SamplingMetadata) -> SamplerOutput:
+        if not sampling_metadata.all_greedy:
+            raise NotImplementedError(
+                "Only greedy sampling is supported for now.")
```

Comment:
Could we just speculate on the greedy sequences in the batch and treat the non-greedy ones normally? If not, I think we should disable spec decoding while there are any non-greedy reqs in the batch rather than failing here.

Is the plan to support random sampling too?

Reply:
I think it is OK to not include it in this PR, as spec decoding feature is experimental at the moment.

Not sure we can support random sampling for N-gram method though. cc @LiuXiaoxuanPKU

Reply:
OK thanks, I had been assuming we might be enabling ngram spec decoding by default but if not yet then this sounds ok.